### ISSUE TITLE
gtag error! unrecognized option --skip-symlink 

### ISSUE THREAD

**Issue by vewe-richard** [Yggdroot/LeaderF#706]
- vim or neovim? 
    - [x] vim
    - [ ] neovim
- Output of `vim --version` or `nvim --version`: VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Oct 10 2020 14:57:00)
- Output of `:echo has("python")`: 0
- Output of `:echo has("python3")`: 1
- Output of `:echo &pythondll`(only vim, not neovim): 
- Output of `:echo &pythonthreedll`(only vim, not neovim): 
- Output of `:py print(sys.version)`: 
- Output of `:py3 print(sys.version)`:  3.6.9
- Output of `:echo g:Lf_Debug_Cmd`: Undefined variable
- Output of `:echo g:Lf_FilesFromCache`:  Undefined variable
- Operating system: 
    - [x] Linux
    - [ ] Mac OS X
    - [ ] Windows
    - [ ] Etc.
- Configurations related to LeaderF in vimrc: 
" don't show the help in normal mode
let g:Lf_HideHelp = 1
let g:Lf_UseCache = 0
let g:Lf_UseVersionControlTool = 0
let g:Lf_IgnoreCurrentBufferName = 1
" popup mode
let g:Lf_WindowPosition = 'popup'
let g:Lf_PreviewInPopup = 1
let g:Lf_StlSeparator = { 'left': "\ue0b0", 'right': "\ue0b2", 'font': "DejaVu Sans Mono for Powerline" }
let g:Lf_PreviewResult = {'Function': 0, 'BufTag': 0 }

let g:Lf_ShortcutF = "<leader>ff"
noremap <leader>fb :<C-U><C-R>=printf("Leaderf buffer %s", "")<CR><CR>
noremap <leader>fm :<C-U><C-R>=printf("Leaderf mru %s", "")<CR><CR>
noremap <leader>ft :<C-U><C-R>=printf("Leaderf bufTag %s", "")<CR><CR>
noremap <leader>fl :<C-U><C-R>=printf("Leaderf line %s", "")<CR><CR>

noremap <C-B> :<C-U><C-R>=printf("Leaderf! rg --current-buffer -e %s ", expand("<cword>"))<CR>
noremap <C-F> :<C-U><C-R>=printf("Leaderf! rg -e %s ", expand("<cword>"))<CR>
" search visually selected text literally
"xnoremap gf :<C-U><C-R>=printf("Leaderf! rg -F -e %s ", leaderf#Rg#visual())<CR>
noremap go :<C-U>Leaderf! rg --recall<CR>

" Do not enable leaderf gtags function
" should use `Leaderf gtags --update` first
let g:Lf_GtagsAutoGenerate = 1
let g:Lf_Gtagslabel = 'native-pygments'
" -r: see reference
noremap <leader>fr :<C-U><C-R>=printf("Leaderf! gtags -r %s --auto-jump", expand("<cword>"))<CR><CR>
" -d: see definiton
noremap <leader>fd :<C-U><C-R>=printf("Leaderf! gtags -d %s --auto-jump", expand("<cword>"))<CR><CR>
noremap <leader>fo :<C-U><C-R>=printf("Leaderf! gtags --recall %s", "")<CR><CR>
noremap <leader>fn :<C-U><C-R>=printf("Leaderf gtags --next %s", "")<CR><CR>
noremap <leader>fp :<C-U><C-R>=printf("Leaderf gtags --previous %s", "")<CR><CR>

### Describe your question, feature request, or bug.
After `let g:Lf_GtagsAutoGenerate = 1`, when I open vim, I see "gtags error!" on the status bar.
Then I enable vim verbose message, and found "--skip-symlink" is not supported by gtags.
The global version is 6.6.2
I can fix it on python/leaderf/gtagsExpl.py, can you assign this issue to me.
I am looking forward to make my first contribute to opensource community.

### Steps to reproduce
vimrc: let g:Lf_GtagsAutoGenerate = 1
open vim

### Actual behaviour
LeaderF gtags command does not work

### Expected behaviour
gtags should work

**Comment by Yggdroot** [comment]
> I can fix it on python/leaderf/gtagsExpl.py, can you assign this issue to me.

No need to assign, if you'd like to contribute to LeaderF, just fork it and open a pull request.

BTW, how are you going to fix it?

**Comment by vewe-richard** [comment]
From gtags/global document, --skip-symlink is a new option after 6.6.3.
My idea is to check the version before using --skip-symlink, if we can read a correct version number and is lower than 6.6.3, then remove --skip-synlink option, except following previous procedure.

Below is my patch, and is tested with different gtags version.

```
richard@richard-NH50-70RA:~/.vim/plugged/LeaderF$ git diff
diff --git a/autoload/leaderf/python/leaderf/gtagsExpl.py b/autoload/leaderf/python/leaderf/gtagsExpl.py
index fbffa29..9a93ea3 100644
--- a/autoload/leaderf/python/leaderf/gtagsExpl.py
+++ b/autoload/leaderf/python/leaderf/gtagsExpl.py
@@ -554,8 +554,22 @@ class GtagsExplorer(Explorer):
         """
         self._accept_dotfiles = "--accept-dotfiles " if lfEval("get(g:, 'Lf_GtagsAcceptDotfiles', '0')") == '1' else ""
         self._skip_unreadable = "--skip-unreadable " if lfEval("get(g:, 'Lf_GtagsSkipUnreadable', '0')") == '1' else ""
-        self._skip_symlink = "--skip-symlink%s " % ('=' + lfEval("get(g:, 'Lf_GtagsSkipSymlink', '')")
+
+        # Fix issue #706, gtags error! unrecognized option --skip-symlink
+        # --skip-symlink option is supported after global 6.6.3
+        proc = subprocess.Popen(["gtags", "--version"], stdout=subprocess.PIPE)
+        try:
+            version = proc.stdout.readline().decode().split()[-1].split(".")
+            # change version string (for example 6.6.2) to a decimal number
+            vernum = int(version[0]) * 10000 + int(version[1]) * 100 + int(version[2])
+            if vernum < 60603:  # if version is less than 6.6.3
+                self._skip_symlink = ""
+            else:
+                raise Exception
+        except Exception as e:  # if we can not parse the versin number, or version number >= 6.6.3
+            self._skip_symlink = "--skip-symlink%s " % ('=' + lfEval("get(g:, 'Lf_GtagsSkipSymlink', '')")
                                 if lfEval("get(g:, 'Lf_GtagsSkipSymlink', '')") != '' else "")
+
         self._gtagsconf = lfEval("get(g:, 'Lf_Gtagsconf', '')")
         if self._gtagsconf:
             self._gtagsconf = self._gtagsconf.join('""')

```

**Comment by Yggdroot** [comment]
`subprocess.Popen` is creating a process, it's a heavy operation, I don't  think it is a good idea to take so much overhead to get the version number, that is why I did not add the version check.

I'd like to add a global variable to tell the version number, if the user is using 6.6.2, he can set the variable to 6.6.2 to solve the issue.

**Comment by vewe-richard** [comment]
I see.
But, how about to read gtags' version on init of GtagsExplorer, to avoid read it every time?

I tested the time cost on reading gtags' version is about 1ms.

```
    t1 = time.time()
    proc = subprocess.run(["gtags", "--version"], stdout=subprocess.PIPE)
    version = proc.stdout.decode().splitlines()[0]
    t2 = time.time()
    print(t2 - t1)
    print(version)
#output 
0.0010597705841064453
gtags (GNU GLOBAL) 6.6.2
```

**Comment by Yggdroot** [comment]
I got an error if using python2:
```
proc = subprocess.run(["gtags", "--version"], stdout=subprocess.PIPE)
AttributeError: 'module' object has no attribute 'run'
```

It takes 0.03889584541320801 second on Windows. That is OK, it's acceptable.

it's better to set `shell=True` when using subprocess on Windows.

**Comment by vewe-richard** [comment]
It depends on whether "gtags" had been loaded into memory?
```
$ python3 scratch.py 
0.0008754730224609375
gtags (GNU GLOBAL) 6.6.5
```
OK, nice to get your approval.

**Comment by vewe-richard** [comment]
Can you have a look again? I did not test it yet.

```
diff --git a/autoload/leaderf/python/leaderf/gtagsExpl.py b/autoload/leaderf/python/leaderf/gtagsExpl.py
index cf7f227..53d3fb6 100644
--- a/autoload/leaderf/python/leaderf/gtagsExpl.py
+++ b/autoload/leaderf/python/leaderf/gtagsExpl.py
@@ -38,6 +38,7 @@ class GtagsExplorer(Explorer):
         self._gtagslibpath = []
         self._result_format = None
         self._last_result_format = None
+        self._gtags_version = self._getGtagsVersion()
         self._evalVimVar()
         self._has_nvim = lfEval("has('nvim')") == '1'
         self._db_timestamp = 0
@@ -556,6 +557,15 @@ class GtagsExplorer(Explorer):
         self._skip_unreadable = "--skip-unreadable " if lfEval("get(g:, 'Lf_GtagsSkipUnreadable', '0')") == '1' else ""
         self._skip_symlink = "--skip-symlink%s " % ('=' + lfEval("get(g:, 'Lf_GtagsSkipSymlink', '')")
                                 if lfEval("get(g:, 'Lf_GtagsSkipSymlink', '')") != '' else "")
+        # Fix issue #706, gtags error! unrecognized option --skip-symlink
+        # --skip-symlink option is only supported after global 6.6.3
+        # remove --skip-symlink option in case the version is lower then 6.6.3
+        try:
+            if self._compare_gtags_version("6.6.3") < 0:
+                self._skip_symlink = ""
+        except:
+            pass
+
         self._gtagsconf = lfEval("get(g:, 'Lf_Gtagsconf', '')")
         if self._gtagsconf:
             self._gtagsconf = self._gtagsconf.join('""')
@@ -884,6 +894,30 @@ class GtagsExplorer(Explorer):
     def getLastResultFormat(self):
         return self._last_result_format
 
+    def _getGtagsVersion(self):
+        proc = subprocess.Popen(["gtags", "--version"], shell=True, stdout=subprocess.PIPE)
+        try:
+            version = proc.stdout.readline().decode().split()[-1].split(".")
+            return int(version[0]) * 10000 + int(version[1]) * 100 + int(version[2])
+        except:
+            return None
+
+    def _compare_gtags_version(self, specify_version):
+        """
+        comparing current gtags version with the specified version string
+        Args:
+            specify_version: the specified version string, for example 6.6.3
+        return:
+            >0, greater than the specify_version
+            0, the same version
+            <0, lower than the specify_version
+        """
+        try:
+            specify_vernum = int(specify_version[0]) * 10000 + int(specify_version[1]) * 100 + int(specify_version[2])
+            return self._gtags_version - specify_vernum
+        except:
+            raise Exception("can not compare gtags version")
+
 #*****************************************************
 # GtagsExplManager
 #*****************************************************
```
